src/looplets/jumpers.jl

```diff
@@ -31,61 +52,89 @@ function lower(root::FinchNode, ctx::AbstractCompiler,  style::JumperStyle)
 
-    body_2 = Rewrite(Postwalk(@rule access(~tns::isvirtual, ~mode, ~idxs...) => begin
+    foreach(filter(isvirtual, collect(PostOrderDFS(root.body)))) do node
+        push!(ctx.code.preamble, jumper_seek(node.val, ctx, root.ext))
+    end
+
+    body_1 = Rewrite(Postwalk(@rule access(~tns::isvirtual, ~mode, ~idxs...) => begin
         tns_2 = jumper_body(tns.val, ctx, root.ext)
         access(tns_2, mode, idxs...)
     end))(root.body)
 
-    body_3 = contain(ctx) do ctx_2
+    body_2 = contain(ctx) do ctx_2
         push!(ctx_2.code.preamble, :($i0 = $i))
-        if is_continuous_extent(root.ext) 
-            ctx_2(loop(root.idx, bound_measure_below!(ContinuousExtent(start = value(i0), stop = getstop(root.ext)), literal(0)), body_2))
+        i1 = freshen(ctx_2.code, i)
+
+        ext_1 = bound_measure_below!(similar_extent(root.ext, value(i0), getstop(root.ext)), get_smallest_measure(root.ext))
+        ext_2 = mapreduce((node)->phase_range(node, ctx_2, ext_1), (a, b) -> virtual_union(ctx_2, a, b), PostOrderDFS(body_1))
+        ext_3 = virtual_intersect(ctx_2, ext_1, ext_2)
+        ext_4 = cache_dim!(ctx_2, :phase, ext_3)
+
+        body = Rewrite(Postwalk(node->phase_body(node, ctx_2, ext_1, ext_4)))(body_1)
+        body = quote
+            $i1 = $i
+            $(contain(ctx_2) do ctx_3
+                (ctx_3)(loop(
+                    root.idx,
+                    ext_4,
+                    body
+                ))
+            end)
+            
+            $i = $(ctx_2(getstop(ext_4))) + $(ctx_2(getunit(ext_4)))
+        end
+
+        if query(call(>=, measure(ext_4), 0), ctx_2)  
+            body
         else
-            ctx_2(loop(root.idx, bound_measure_below!(Extent(start = value(i0), stop = getstop(root.ext)), literal(1)), body_2))
+            quote
+                if $(ctx_2(getstop(ext_4))) >= $(ctx_2(getstart(ext_4)))
+                    $body
+                end
+            end
         end
+
     end
 
     @assert isvirtual(root.ext)
 
-    target = is_continuous_extent(root.ext.val) ? 0 : 1
-    if query(call(==, measure(root.ext.val), target), ctx)
-        body_3
+    if query(call(==, measure(root.ext.val), get_smallest_measure(root.ext.val)), ctx)
+        body_2
     else
         return quote
             while $guard
-                $body_3
+                $body_2
             end
         end
     end
 end
 
-jumper_body(node::Jumper, ctx, ext) = node.body
-jumper_body(node, ctx, ext) = node
-
-@kwdef struct Jump
-    seek = nothing
-    stop
-    body
-    next = nothing
-end
-
-FinchNotation.finch_leaf(x::Jump) = virtual(x)
-
-(ctx::Stylize{<:AbstractCompiler})(node::Jump) = ctx.root.kind === loop ? JumperPhaseStyle() : DefaultStyle()
-
-function phase_range(node::Jump, ctx, ext)
-    push!(ctx.code.preamble, node.seek !== nothing ? node.seek(ctx, ext) : quote end)
-    similar_extent(ext, getstart(ext), node.stop(ctx, ext))
+function phase_range(node::AcceptJumper, ctx, ext)
```

Comment:
I noticed you introduced phase_range and phase_body for AcceptJumper, but maybe we can just use Jumper_range and jumper_body?